### PR TITLE
feat(abi)!: add explicit return type field to ABI.

### LINKED ISSUE #678: Add an explicit return type field into ABI
# Problem

Currently the ABI does not have a concept of a return parameter, only parameters. So to emulate a return parameter, a constant `MAIN_RETURN_NAME` is used in the parameter list.

# Solution

Ad `return_type` to ABI. This means that we can remove the `MAIN_RETURN_NAME` constant from the compiler and it is solely a value that lives in `noirc_abi`. When the Verifier.toml file is parsed, this is the only time that one needs to check for `MAIN_RETURN_NAME` so that it can be placed into the `return_parameter` field.

Note that the field to add is `return_type` and not `return_parameter` :

When converting noir return parameters into ABI parameters; we are currently doing this for only the main function. This is more abundant by the fact that we name that return parameter `MAIN_RETURN_NAME`. We have already noted that the main return type needs to be public, so the visibility field here is not useful. (There is an argument to be made that to make the function more generic, we should return the visibility) For this argument, we can simply return the type and visibility. 

# Alternatives considered

# Additional context

### PR DESCRIPTION
**Pull request by TomAFrench**
# Related issue(s)

Resolves #678 

# Description

## Summary of changes

Draft pending proper error types.

The ABI now has separate fields for the type and witness assignments of the return value. As a know on effect, the circuit's return value is now no longer included in the `InputMap` and so is taken as a separate argument to `Abi.encode` and returned separately by `Abi.decode`.

Some care needs to be taken now when checking to see if an ABI contains a public input as this could either be a public param or a return value, whereas previously we could just search `abi.parameters` for a public parameter. I've added a getter to perform this check for us.

## Test additions / changes

# Checklist

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [x] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [x] I have reviewed the changes on GitHub, line by line.
- [x] I have ensured all changes are covered in the description.
- [ ] This PR requires documentation updates when merged.

# Additional context

### CODE REVIEW

File: crates/noirc_abi/src/lib.rs [hunk 1]
```diff
@@ -208,19 +229,43 @@ impl Abi {
...
+                    |(&witness, &field_element)| match witness_map.insert(witness, field_element) {
+                        Some(existing_value) if existing_value != field_element => {
+                            Err(AbiError::InconsistentWitnessAssignment(witness))
+                        }
+                        _ => Ok(()),
+                    },
+                )?;
+            }
+            (None, Some(_)) => {
+                return Err(AbiError::UnexpectedParams(vec![MAIN_RETURN_NAME.to_owned()]))
```
Comment (TomAFrench) [inline crates/noirc_abi/src/lib.rs:261]: ditto

File: crates/noirc_abi/src/lib.rs [hunk 2]
```diff
@@ -208,19 +229,39 @@ impl Abi {
...
-    }
-
-    /// Checks that no extra witness values have been provided.
-    fn check_for_unexpected_inputs(&self, inputs: &InputMap) -> Result<(), AbiError> {
-        let param_names = self.parameter_names();
-        if param_names.len() < inputs.len() {
-            let unexpected_params: Vec<String> =
-                inputs.keys().filter(|param| !param_names.contains(param)).cloned().collect();
-            return Err(AbiError::UnexpectedParams(unexpected_params));
+        // The user can optionally provide a return value to be inserted into the witness map.
```
Comment (kevaundray) [inline crates/noirc_abi/src/lib.rs:232]: How does the user optionally provide a return value from the toml file?

Comment (TomAFrench) [reply crates/noirc_abi/src/lib.rs:232]: This may be getting some wires crossed due to https://github.com/noir-lang/noir/pull/867. This optional return value is to differentiate between encoding the partial witness for executing the program (no return value passed) or encoding the public inputs to pass to the verifier (potentially with return value passed). It's not optional in the toml file however, it either needs to be there or can't exist.

I can reword this to be a bit clearer though.

Comment (kevaundray) [reply crates/noirc_abi/src/lib.rs]: yeah I had just finished reviewing that so the word "user" was getting my wires crossed

**Review by kevaundray — APPROVED** [review]
